### ISSUE TITLE
Stuttter lint

### ISSUE THREAD

**Issue by bbatha** [Manishearth/rust-clippy#904]
Clippy should lint for stuttered names. For instance,

```rust
    mod foo {
            struct FooBar {}
    }
```

Should warn the that users will write the type `foo::FooBar` and suggest the the name `foo::Bar` instead. This lint should probably be default allow.

**Comment by mcarton** [comment]
This is kind of like the [`enum_variant_names`](https://github.com/Manishearth/rust-clippy/wiki#enum_variant_names) lint.

**Comment by llogiq** [comment]
There is some overlap, but the stutter lint would also work when there's only a single variant.

**Comment by oli-obk** [comment]
This would trigger in clippy for 

* `enum_clike::EnumClikeUnportableVariant`
* `shadow::ShadowPass`
* `loops::LoopsPass`
* `zero_div_zero::ZeroDivZeroPass`
* `needless_update::NeedlessUpdatePass`
* `no_effect::NoEffectPass`
* `map_clone::MapClonePass`
* `temporary_assignment::TemporaryAssignmentPass`
* `escape::EscapePass`
* `panic::PanicPass`
* `print::PrintLint`
* `drop_ref::DropRefPass`
* `regex::RegexPass`
* `format::FormatMacLint`

Might trigger, depending on the impl:

* `minmax::MinMaxPass`
* `open_options::NonSensicalOpenOptions`
* `vec::UselessVec`

Also there are lots of types whose name is the camel case version of their enclosing module

**Comment by mcarton** [comment]
> Do we really want to name all of these `modulename::Pass`

I wouldn’t mind that but:

> Also there are lots of types whose name is the camel case version of their enclosing module

That’s really common in Rust world indeed:

* [`std::cell::Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html);
* [`std::string::String`](https://doc.rust-lang.org/std/string/struct.String.html);
* [`iron::headers::Headers`](http://ironframework.io/doc/iron/headers/struct.Headers.html);
* [`num::bigint::BigInt`](https://rust-num.github.io/num/num/bigint/struct.BigInt.html);
* etc.

**Comment by oli-obk** [comment]
> That’s really common in Rust world indeed:

Right, but if we start mass-renaming some lints to `modulename::Pass` and leave others as `module_name::ModuleName`, it's somewhat inconsistent
